app/models/manageiq/providers/amazon/inventory/collector/cloud_manager.rb

```diff
@@ -12,25 +12,41 @@ def availability_zones
   end
 
   def key_pairs
-    hash_collection.new(aws_ec2.client.describe_key_pairs[:key_pairs])
+    hash_collection.new(aws_ec2.client.describe_key_pairs[:key_pairs]).all
   end
 
   def private_images
-    hash_collection.new(aws_ec2.client.describe_images(:owners  => [:self],
-                                                       :filters => [{:name   => "image-type",
-                                                                     :values => ["machine"]}])[:images])
+    return [] unless options.get_private_images
+
+    @private_images_hashes ||= hash_collection.new(
+      aws_ec2.client.describe_images(:owners  => [:self],
+                                     :filters => [{:name   => "image-type",
+                                                   :values => ["machine"]}]).images).all
   end
 
   def shared_images
-    hash_collection.new(aws_ec2.client.describe_images(:executable_users => [:self],
-                                                       :filters          => [{:name   => "image-type",
-                                                                              :values => ["machine"]}])[:images])
+    return [] unless options.get_shared_images
+
+    @shared_images_hashes ||= hash_collection.new(
+      aws_ec2.client.describe_images(:executable_users => [:self],
+                                     :filters          => [{:name   => "image-type",
+                                                            :values => ["machine"]}]).images).all
   end
 
   def public_images
-    filters = options.public_images_filters
-    hash_collection.new(aws_ec2.client.describe_images(:executable_users => [:all],
-                                                       :filters          => filters)[:images])
+    return [] unless options.get_public_images
+
+    @public_images_hashes ||= hash_collection.new(
+      aws_ec2.client.describe_images(:executable_users => [:all],
+                                     :filters          => options.public_images_filters).images).all
+  end
+
+  def referenced_images
+    return [] if extra_image_references.blank?
+
+    hash_collection.new(
+      aws_ec2.client.describe_images(:filters => [{:name => 'image-id', :values => extra_image_references}]).images
```

Comment:
you if there is an api limit on how large `:values` can be?
Maybe (in the future) this will need a `.batch`ed version